File: web-dashboard/src/domain/payments/service-demand.ts
// Types & helpers du domaine « prestations les plus demandées » — couche domaine
// (hexagonal, ADR-0008), TypeScript pur, testable sans React. **Parité stricte**
// avec le backend (`coiflink_api/domain/service_demand.py`, US-6.3 #41) : pour un
// salon (et une période optionnelle), les prestations classées **par volume** (RDV
// COMPLETED) et **par revenu** (somme des `price_at_booking` figés), deux ordres et
// les mêmes entrées.
//
// Le backend reste **l'autorité des chiffres ET de l'ordre** : ce module ne fait que
// **projeter** la réponse et **formater** l'affichage — il ne recalcule ni ne
// **re-trie** jamais (invariant #31). Volume est un entier ; revenu est porté en
// **chaîne décimale** (`NUMERIC(12,2)`) pour ne pas perdre de précision via un
// flottant JavaScript. Aucun secret, aucune PII (pas de `client_id`/`appointment_id`).

// Réutilisés tels quels : montant FCFA (payment.ts) et « ×N fois » (customer/stats.ts).
export { formatXof } from "./payment";
export { formatOccurrences } from "@/src/domain/customer/stats";

// Devise unique du MVP (mono-devise XOF/FCFA — PRD §9.6).
export { DEFAULT_CURRENCY } from "./payment";

// Une prestation du classement : libellé + volume (entier) + revenu (chaîne décimale).
export interface ServiceDemandItem {
  serviceId: string;
  name: string;
  // Nombre d'occurrences réalisées (RDV COMPLETED) de cette prestation.
  volume: number;
  // Somme des priceAtBooking (prix figés), chaîne décimale — jamais de flottant.
  revenue: string;
}

// Les prestations du salon dans deux ordres (miroir de `ServiceDemandRanking` du
// backend). Objet-valeur **sans PII** : uniquement des prestations nommées, des
// compteurs, des montants, une période et une devise. `byVolume`/`byRevenue`
// portent **les mêmes** entrées, déjà triées côté serveur.
export interface ServiceDemandRanking {
  currency: string; // «XOF» au MVP
  // Bornes ISO "YYYY-MM-DD" (Africa/Abidjan) ; `null` = toute l'histoire.
  dateFrom: string | null;
  dateTo: string | null;
  byVolume: ServiceDemandItem[];
  byRevenue: ServiceDemandItem[];
}

// Les deux axes de classement proposés à l'AC US-6.3 (« par volume et par revenu »).
export type ServiceDemandMetric = "volume" | "revenue";

// Formate une période en légende compacte `JJ/MM/AAAA → JJ/MM/AAAA`. Absence de
// bornes (toute l'histoire) → libellé neutre. Présentation uniquement ; jamais de
// recalcul de la donnée.
export function formatDemandPeriod(ranking: ServiceDemandRanking): string {
  const { dateFrom, dateTo } = ranking;
  if (!dateFrom && !dateTo) return "Depuis l'ouverture";
  if (dateFrom && dateTo) {
    if (dateFrom === dateTo) return formatIsoDateFr(dateFrom);
    return `${formatIsoDateFr(dateFrom)} → ${formatIsoDateFr(dateTo)}`;
  }
  if (dateFrom) return `À partir du ${formatIsoDateFr(dateFrom)}`;
  return `Jusqu'au ${formatIsoDateFr(dateTo as string)}`;
}

// Formate une date ISO "YYYY-MM-DD" en `JJ/MM/AAAA` (fr-FR), sans dépendre du fuseau
// du navigateur (découpage textuel, pas de `new Date()`). Chaîne mal formée renvoyée
// telle quelle (défensif).
function formatIsoDateFr(iso: string): string {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(iso);
  if (!match) return iso;
  const [, year, month, day] = match;
  return `${day}/${month}/${year}`;
}

File: backend/coiflink_api/domain/service_demand.py
"""Prestations les plus demandées d'un salon (domaine pur, US-6.3, #41).

Ce module porte les objets-valeur de lecture du classement (`ServiceDemand`,
`ServiceDemandRanking`) et — surtout — la **fonction pure de tri**
(`rank_service_demand`) qui **est** la règle métier de l'US-6.3 : « classer les
prestations par volume **et** par revenu généré » se traduit par la définition
**exacte** de deux ordres déterministes (par volume décroissant, par revenu
décroissant) sur les mêmes prestations agrégées. L'isoler ici (sans I/O) le rend
testable sans base et garantit un ordre **stable**, indépendant de l'ordre SQL.

Conformément à l'hexagonal (ADR-0008), le module ne connaît ni FastAPI ni
SQLAlchemy : l'adapter sortant (`adapters/outbound/persistence/queue_ticket_repository
.py::demand_by_service`) fait l'**agrégat en base** (`GROUP BY service_id`, `COUNT`
volume, `SUM(services.price)` revenu **courant**) et renvoie des `ServiceDemand`
**non triés** ; le cas d'usage (`application/service_demand.py`) les ordonne via
`rank_service_demand` ; l'adapter entrant (`adapters/inbound/stats.py`) projette le
résultat en JSON.

**Volume & revenu = tickets `done` uniquement** (réalisés), en cohérence avec
l'invariant §8.1. Un ticket `waiting`/`called`/`in_progress`/`expired` ne pèse
**ni** en volume **ni** en revenu — « non réalisés exclus » (§8.1) est vrai **par
construction** du filtre de statut décidé côté serveur.

**Revenu = somme des `services.price` courants** (pivot walk-in : `queue_ticket_
services` ne fige aucun prix, contrairement à l'ancien `appointment_services
.price_at_booking` — décision « résolution en direct » #148), devise **XOF**
(§9.6), `Decimal` de bout en bout (`NUMERIC(12,2)`, sérialisé en **chaîne**
décimale, jamais de flottant). Cette grandeur (valeur des prestations réalisées)
peut **différer** du CA du salon (#40, dérivé du journal de caisse net).
"""

from __future__ import annotations

import datetime
import decimal
import uuid
from dataclasses import dataclass

from coiflink_api.domain.payment import DEFAULT_CURRENCY


@dataclass(frozen=True)
class ServiceDemand:
    """Une prestation dans le classement du salon (US-6.3, #41).

    `volume` = nombre d'occurrences **réalisées** (une ligne `appointment_services`
    par occurrence, RDV `COMPLETED`) ; `revenue` = somme des `price_at_booking` de
    cette prestation (prix **figés**, XOF, jamais le tarif courant). `name` = libellé
    **courant** (`services.name`), résoluble même si la prestation est soft-deletée
    (`is_active = false`, FK `RESTRICT`). Clé métier = `service_id` (deux prestations
    distinctes partageant un libellé ne sont pas fusionnées). `Decimal` de bout en
    bout, jamais de flottant. Objet-valeur **immuable** et **sans PII** (§11.3) :
    aucun `client_id`, `appointment_id`, ni ligne de RDV/paiement.
    """

    service_id: uuid.UUID
    name: str
    volume: int
    revenue: decimal.Decimal


@dataclass(frozen=True)
class ServiceDemandRanking:
    """Classement des prestations d'un salon : mêmes entrées, deux ordres (US-6.3).

    `by_volume` = tri **volume décroissant** ; `by_revenue` = tri **revenu
    décroissant**. Chaque entrée porte **les deux** métriques (le web affiche « ×N
    fois · M FCFA » quel que soit l'onglet). `date_from`/`date_to` échoient la période
    couverte (`None` si toute l'histoire). Classements **vides** si aucun RDV réalisé
    (état normal, pas une erreur). Immuable et **sans PII** : seulement des
    prestations nommées, des compteurs, des montants, une période et une devise.
    """

    by_volume: tuple[ServiceDemand, ...] = ()
    by_revenue: tuple[ServiceDemand, ...] = ()
    date_from: datetime.date | None = None
    date_to: datetime.date | None = None
    currency: str = DEFAULT_CURRENCY


def rank_service_demand(
    rows: tuple[ServiceDemand, ...],
    *,
    date_from: datetime.date | None = None,
    date_to: datetime.date | None = None,
    currency: str = DEFAULT_CURRENCY,
) -> ServiceDemandRanking:
    """Trie les prestations agrégées en deux classements déterministes (fonction pure).

    `by_volume` : `-volume`, puis `-revenue`, puis `name` croissant, puis
    `str(service_id)` croissant. `by_revenue` : `-revenue`, puis `-volume`, puis
    `name` croissant, puis `str(service_id)` croissant. Les départages sont
    **stables** (indépendants de l'ordre d'itération SQL, utiles aux tests). Les
    `rows` proviennent d'un `GROUP BY service_id` (déjà dédupliquées par prestation) :
    cette fonction **ne ré-agrège pas**, elle **ordonne**. Une entrée vide donne deux
    classements vides. `Decimal` conservé (aucun arrondi, aucun flottant).
    """

    by_volume = tuple(
        sorted(
            rows,
            key=lambda entry: (
                -entry.volume,
                -entry.revenue,
                entry.name,
                str(entry.service_id),
            ),
        )
    )
    by_revenue = tuple(
        sorted(
            rows,
            key=lambda entry: (
                -entry.revenue,
                -entry.volume,
                entry.name,
                str(entry.service_id),
            ),
        )
    )
    return ServiceDemandRanking(
        by_volume=by_volume,
        by_revenue=by_revenue,
        date_from=date_from,
        date_to=date_to,
        currency=currency,
    )


__all__ = [
    "ServiceDemand",
    "ServiceDemandRanking",
    "rank_service_demand",
]
